README.md
# homework
Бот для хранения домашнего задания в Telegram!

Основные библиотеки: aiogram, aiosqlite, logging

# Команды
- /help - помощь по командам(1+ уровень доступа)
- /dz - просмотр д/з(1+ уровень доступа)


- /dz_new *код предмета* *домашнее задание* - обновить д/з(2+ уровень доступа)
- /dz_clear *код предмета* - очистить д/з(2+ уровень доступа)


- /dz_setadmin *id пользователя* *уровень доступа* - назначить доступ пользователю(3+ уровень доступа)

# Настройка
Основная настройка находится в файле **configs/config.py**

- token - токент бота от BotFather
- logs_group - id группы, где расположены логи

С помощью приложения **DB Browser (SQLite)** вы можете выдать себе уровень доступа в самом начале

# Настройка предметов
Вся настройка предметов находится в файле **configs/lessons_config.py**

- lessons - запись предметов в формате "код предмета в базе данных" : "название предмета в именительном падеже"
- lessons_dat - запись предметов в формате "код предмета в базе данных" : "название предмета в дательном падеже"


information/changelog.txt
v1.0
РЕЛИЗ

v1.1
- Исправление ошибки при несуществовании пользователя во всех командах и функциях

v2.0
- Добавил глоабальную удобнейшую систему редактирования предметов
- Теперь в /start имя и фамилия получаются от самого юзера, а не "Не установлено"
- Общая оптимизация кода и исправление багов и недоработок

v2.1
- Теперь при авторизации пишет имя пользователя
- Запрет на введение /dz в группах
- Введение логов


v3.0
- Полностью переписана система, доработана до новейшей версии aiogram
- Изменено сообщение, которое пишется в логи
- Убран префикс "dz_" перед командой /help
- Добавлена новая настройкав configs/config.py - запрет на отправку /dz в группы (по умолчанию True)
- Исправлена ошибка, при которой можно было получить домашку по кнопке из старого сообщения
- Исправлена проблему со слишком длинным д/з
- Исправлен баг с ненахождением предметов, коды которых заканчиваются и начинаются на буквы c h e k
- Исправлено множество других багов

main.py
import logging

from asyncio import run
from datetime import datetime

from aiogram import Dispatcher, types, Bot, Router, F
from aiogram.filters import Command, CommandObject
from aiogram.utils.markdown import hlink

from database import Database
from configs.config import token, logs_group, anti_group_dz
from configs.lessons_config import lessons
from utils.utils import get_text_list_lessons_with_code

#Настройки бота
bot = Bot(token, parse_mode='HTML')
dp = Dispatcher(disable_fsm=True)
router = Router()
logging.basicConfig(level=logging.INFO)
db = Database('domashka.db')


# Обработка запросов на получение домашки
@router.callback_query(F.data.startswith('check_'))
async def callback_check_dz(call: types.CallbackQuery):
    dostup = await db.get_user_dostup(call.from_user.id)
    if dostup is None or dostup < 1: return await call.answer("😢 У вас нет 1 уровня доступа!")

    lesson_code = call.data.split('_', maxsplit=1)[1]    
    data = await db.get_dz(lesson_code)

    if data is None or data[0] is None: return await call.answer(f"📔 Домашка по {lessons[lesson_code][1]} отсутствует!", show_alert=True)    
    await call.message.answer(f'📔 Доска д/з по {lessons[lesson_code][1]}\n🕓 Время добавления: {data[2]} ({hlink("автор", data[1])})\n\n📚 Задание:\n{data[0]}', parse_mode='HTML')
    return call.answer()

#///////////////////////////////////////////////////

@dp.message(Command('start'))
async def start(message: types.Message):
    registration = await db.add_user((message.from_user.id, message.from_user.first_name, message.from_user.last_name))
    
    if registration == 'Error': return print('ОШИБКА В ЗАПРОСЕ!!!!')
    if not registration: return await message.reply(f"🔑 Успешная авторизация! Здравствуйте, {message.from_user.first_name}")
    return await message.reply(f"🎉 Вы успешно зарегистрировались в боте! Ваш ID телеграм - <b>{message.from_user.id}</b>")


@dp.message(Command('dz'))
async def check_dz(message: types.Message):
    dostup = await db.get_user_dostup(message.from_user.id)
    if dostup is None or dostup < 1: return await message.reply("😢 У вас нет 1 уровня доступа!")
    if message.chat.id < 0 and anti_group_dz: return await message.reply("😢 В целях анти-флуда данную команду <b>НЕЛЬЗЯ</b> использовать в группах!\nПожалуйста, используйте ее в личке с ботом")
    
    keyboard = types.InlineKeyboardMarkup(inline_keyboard = [[types.InlineKeyboardButton(text=lessons[i][0], callback_data=f'check_{i}')] for i in lessons])
    return await message.reply('🛍 Выберите *школьный предмет* из списка ниже', parse_mode= "Markdown", reply_markup=keyboard)


@dp.message(Command('dz_new'))
async def new_dz(message: types.Message, command: CommandObject):
    dostup = await db.get_user_dostup(message.from_user.id)
    if dostup is None or dostup < 2: return await message.reply("😢 У вас нет 2 уровня доступа!")

    if command.args is None: return await message.reply(f"💻 Для добавления домашки используйте следующий синтакисис:\n<b>/dz_new *предмет* *домашка*</b>\n\nПример:\n<b>/dz_new en1 1. Учебник страница 51 упражнение устно, рабочая тетрадь стр 31 полностью</b>\n\nДоступные предметы:\n<b>{get_text_list_lessons_with_code()}</b>")
    info = command.args.split(' ', maxsplit=1)
    if len(info) < 2: return await message.reply(f"💻 Для добавления домашки используйте следующий синтакисис:\n<b>/dz_new *предмет* *домашка*</b>\n\nПример:\n<b>/dz_new en1 1. Учебник страница 51 упражнение устно, рабочая тетрадь стр 31 полностью</b>\n\nДоступные предметы:\n<b>{get_text_list_lessons_with_code()}</b>")
    if info[0] not in lessons: return await message.reply(f"❗️ Вы указали неверный код предмета!\n\nДоступные предметы:\n{get_text_list_lessons_with_code()}")
    
    await db.add_new_dz(message.from_user.id, message.from_user.url, info[0], info[1])
    await message.reply(f"☑️ Вы успешно изменили домашку по <b>{lessons[info[0]][1]}</b> на:\n\n<b>{info[1]}</b>")
    return await bot.send_message(logs_group, f"🤨 Пользователь: {hlink(message.from_user.first_name, message.from_user.url)}\nДействие: <b>изменение домашки по {lessons[info[0]][1]}</b>\nНовая домашка: <b>{info[1]}</b>")


@dp.message(Command('help'))
async def help_cmd(message: types.Message):
    dostup = await db.get_user_dostup(message.from_user.id)
    if dostup is None or dostup < 1: return await message.reply("😢 У вас нет 1 уровня доступа!")
    return await message.reply("😍Команды ДомашкоБота:\n\n<b>/dz</b> - узнать домашнее задание (1+ уровень доступа)\n<b>/dz_help</b> - открыть меню помощи (1+ уровень доступа)\n<b>/dz_new *код предмета* *домашнее задание*</b> - обновить домашнее задание (2+ уровень доступа)\n<b>/dz_clear *код предмета*</b> - очистить домашнее задание (2+ уровень доступа)\n/dz_setadmin *id пользователя* *уровень доступа*")


@dp.message(Command('dz_clear'))
async def clear_cmd(message: types.Message, command: CommandObject):
    dostup = await db.get_user_dostup(message.from_user.id)
    if dostup is None or dostup < 2: return await message.reply("😢 У вас нет 2 уровня доступа!")

    info = command.args
    if info is None: return await message.reply(f"💻 Для очистки домашки используйте следующий синтакисис:\n<b>/dz_clear предмет</b>\n\nПример:\n<b>/dz_clear en1</b>\n\nДоступные предметы:\n<b>{get_text_list_lessons_with_code()}</b>")
    if info not in lessons: return await message.reply(f"❗️ Вы указали неверный код предмета!\n\nДоступные предметы:\n<b>{get_text_list_lessons_with_code()}</b>")

    await db.clear_dz(message.from_user.id, info)
    await message.reply(f"☑️ Вы очистили домашку по <b>{lessons[info][1]}</b>")
    return await bot.send_message(logs_group, f"🤨 Пользователь: {hlink(message.from_user.first_name, message.from_user.url)}\nДействие: <b>обнуление домашки по {lessons[info][1]}</b>\nРезультат: <b>успешно</b>")


@dp.message(Command('dz_setadmin'))
async def clear_cmd(message: types.Message, command: CommandObject):
    dostup = await db.get_user_dostup(message.from_user.id)
    if dostup is None or dostup < 3: return await message.reply("😢 У вас нет 3 уровня доступа!")

    if command.args is None: return await message.reply("❗️ Неверный синтаксис!\n\nИспользуйте: <b>/dz_setadmin id lvl</b>\nПример: <b>/dz_setadmin 1231231234 2</b>")
    info = command.args.split(maxsplit=1)
    if len(info) < 2: return await message.reply("❗️ Неверный синтаксис!\n\nИспользуйте: <b>/dz_setadmin id lvl</b>\nПример: <b>/dz_setadmin 1231231234 2</b>")

    if await db.get_user_dostup(message.from_user.id) is None: return await message.reply("❗️ Пользователь не зарегистрирован в системе бота!")

    try: 
        if int(info[1]) < 0 and int(info[1]) > 3: return await message.reply("❗️ Уровень должен быть больше или равен 0, но меньше или равен 3!")
    except ValueError:
        return await message.reply("❗️ В аргументах команды нельзя использовать буквы!")
        
    await db.set_user_dostup(message.from_user.id, int(info[1]))
    await message.reply(f"😁 Вы успешно назначили пользователю с ID <b>{info[0]}</b> доступ уровня <b>{info[1]}</b>!")
    return await bot.send_message(int(info[0]), f"😁 Вам успешно выдали доступ уровня <b>{info[1]}</b>!")


# Главный поток
async def main():
    await db.check_db()
    dp.include_router(router)
    print(f"{datetime.now()} | ДомашкоБот v. 3.0 успешно запущен!")
    return await dp.start_polling(bot)


if __name__ == '__main__':
    run(main())


database.py
import aiosqlite
from datetime import datetime
from configs.lessons_config import lessons

class Database:
    def __init__(self, database_name: str):
        self.database_name = database_name


    async def check_db(self):
        """
        Проверяет таблицу на наличие строчки с предметом

        RETURN:
        :new_lessons - список текущих предметов
        """

        async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
            old_lessons = await db.execute('SELECT predmet FROM domashka')
            old_lessons = await old_lessons.fetchall()

            new_lessons = []
            for i in old_lessons:
                if i[0] not in lessons:
                    await db.execute('DELETE FROM domashka WHERE predmet = ?', (i[0],))
                    await db.commit()
                    continue
                new_lessons.append(i[0])

            for i in lessons:
                if i not in new_lessons:
                    await db.execute("INSERT INTO domashka(predmet) VALUES(?)", (i,))
                    await db.commit()
                    new_lessons.append(i)

            return new_lessons


    async def add_user(self, user_info: tuple):
        """
        Добавляет кортеж из данных пользователя (user_info) в базу данных

        RETURN:
        :user_info, в случае успеха (id, first_name, last_name)
        :False, в случае ошибки

        :param user_info - Кортеж из следующей информации: id пользователя, username, уровень админки, ник в боте, количество варнов, уровень VIP, общее кол-во exp, кол-во exp необходимые до нового уровня, кол-во exp собранных для нового лвл, уровень
        """

        try:
            async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
                data = await db.execute(f"SELECT id FROM users WHERE id = {user_info[0]}")

                if await data.fetchone() is None:
                    await db.execute("INSERT INTO users(id, name, surname, dostup) VALUES(?, ?, ?, ?)", (user_info[0], user_info[1], user_info[2], 0))
                    await db.commit()
                    return user_info
                return False
        
        except aiosqlite.OperationalError:
            return 'Error'
        

    async def get_user_dostup(self, user_id: int):
         """
         Получить уровень доступа пользователя

         RETURN:
         :dostup - уровень доступа пользователя
         :None - пользователь не найден

         :param user_id - id пользователя
         """

         async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
                data = await db.execute(f"SELECT dostup FROM users WHERE id = ?", (user_id,))
                if data is None: return None
                data = await data.fetchone()
                return int(data[0])
         

    async def set_user_dostup(self, user_id: int, dostup: int):
        """
        Именить уровень доступа пользователя

         RETURN:
         :dostup - новый уровень доступа пользователя

         :param user_id - id пользователя
         :param dostup - новый уровень доступа
         """
        
        async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
            await db.execute(f"UPDATE users SET dostup = ? WHERE id = ?", (dostup, user_id))
            await db.commit()
            return dostup
         

    async def add_new_dz(self, author_id: int, author_link: str, lesson_code: str, text: str):
        """
        Изменить домашнее задание по предмету

        RETURN:
        :True - в любом случае

        :param author_id - id автора изменения домашки
        :param author_link - url автора изменения домашки
        :param lesson_code - код предмета
        :param text - новая домашка
        """

        async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
            await db.execute(f"INSERT INTO logs(user_id, action, result) VALUES(?, ?, ?)", (author_id, f'изменение д/з {lesson_code}', text))
            await db.execute(f"UPDATE domashka SET dz = ?, author = ?, time = ? WHERE predmet = ?", (text, author_link, datetime.now().strftime('%d.%m.%Y %H:%M:%S'), lesson_code))
            await db.commit()
            return True


    async def clear_dz(self, author_id: int, lesson_code: str):
        """
        Очистить домашнее задание по предмету

        RETURN:
        :True - в любом случае

        :param author_id - id автора очистки
        :param lesson_code - код предмета
        """
        
        async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
            await db.execute(f"UPDATE domashka SET dz = null, author = null, time = null WHERE predmet = ?", (lesson_code,))
            await db.execute(f"INSERT INTO logs(user_id, action, result) VALUES(?, ?, ?)", (author_id, f'очистка {lesson_code}', 'успешно'))
            await db.commit()
            return True
        

    async def get_dz(self, lesson_code: str):
        """
        Получить информацию о домашке

        RETURN:
        :lesson_info - инфа о домашке

        :param lesson_code - код урока
        """

        async with aiosqlite.connect(self.database_name, check_same_thread=False) as db:
            data = await db.execute(f"SELECT dz, author, time FROM domashka WHERE predmet = ?", (lesson_code,))
            data = await data.fetchone()
            return data


domashka.py
import sqlite3
import datetime
import aiogram
import logging
import configparser

from aiogram import Dispatcher, executor, types
from aiogram.contrib.middlewares.logging import LoggingMiddleware
from aiogram.contrib.fsm_storage.memory import MemoryStorage

datetime.datetime.now().min

db = sqlite3.connect('domashka.db', check_same_thread=False)
c = db.cursor()


config = configparser.ConfigParser()
config.read("settings.ini")

#Настройки бота

logs_group = int(config["Main"]["logs_group"]) #Группа для логов
bot = aiogram.Bot(str(config["Main"]["token"]), parse_mode='HTML')
dp = Dispatcher(bot, storage=MemoryStorage())

#///////////////////////////////////////////////////

dp.middleware.setup(LoggingMiddleware())
logging.basicConfig(level=logging.INFO)

date = datetime.datetime.now()
print(str(date), "|" + " ДомашкоБот v. 1.0 успешно запущен!")

# ключ: предмет в дательном падеже
predmets = {'alg': 'алгебре', 'en1': 'первой группе английского', 'en2': 'второй группе английского', 'bio': 'биологии', 'geom': 'геометрии', 'geog': 'географии',
            'chem': 'химии', 'info1': 'первой группе информатики', 'info2': 'второй группе информатики', 'litra': 'литературе', 'nem1': 'первой группе немецкого',
            'nem2': 'второй группе немецкого', 'obchest': 'обществознание', 'history': 'истории', 'russian': 'русскому языку', 'techn1': 'технологии(мальчики)',
            'techn2': 'технологии(девочки)', 'phisic': 'физике'}


@dp.message_handler(commands=['start'])
async def start(message):
    user_id = message.from_user.id
    c.execute(f"SELECT id FROM users WHERE id = {user_id}")
    data = c.fetchone()

    if data[0] is None:
        dostup = 0
        name = "Не установлено"
        surname = "Не установлено"

        user_info = (int(user_id), str(name), str(surname), int(dostup))
        c.execute("INSERT INTO users(id, name, surname, dostup) VALUES(?, ?, ?, ?)",(user_info))
        db.commit()
        await message.reply("🎉 Вы успешно зарегистрировались в боте! Ваш ID телеграм - *" + str(user_id) + '*', parse_mode= "Markdown")

    else:
        await message.reply("🔑 Успешный вход!")


@dp.message_handler(commands=['dz'])
async def check_dz(message):
    user_id = message.from_user.id
    c.execute(f"SELECT dostup FROM users WHERE id = {user_id}")
    data = c.fetchone()

    if int(data[0]) < 1 or data[0] is None:
        await message.reply("😢 У вас меньше 1 уровня доступа!")

    else:
        kb = types.InlineKeyboardMarkup()
        kb_algebra = types.InlineKeyboardButton(text='Алгебра', callback_data='check_algebra')
        kb_english = types.InlineKeyboardButton(text='Английский язык', callback_data='check_english')
        kb_biology = types.InlineKeyboardButton(text='Биология', callback_data='check_biology')
        kb_geogr = types.InlineKeyboardButton(text='География', callback_data='check_geogr')
        kb_geoma = types.InlineKeyboardButton(text='Геометрия', callback_data='check_geoma')
        kb_chimiya = types.InlineKeyboardButton(text='Химия', callback_data='check_chimiya')
        kb_informat = types.InlineKeyboardButton(text='Информатика', callback_data='check_informat')
        kb_litra = types.InlineKeyboardButton(text='Литература', callback_data='check_litra')
        kb_nemec = types.InlineKeyboardButton(text='Немецкий язык', callback_data='check_nemec')
        kb_obshestvo = types.InlineKeyboardButton(text='Обществознание', callback_data='check_obshestvo')
        kb_history = types.InlineKeyboardButton(text='История', callback_data='check_history')
        kb_russian = types.InlineKeyboardButton(text='Русский язык', callback_data='check_russian')
        kb_techno = types.InlineKeyboardButton(text='Технология', callback_data='check_techno')
        kb_phisic = types.InlineKeyboardButton(text='Физика', callback_data='check_phisic')

        kb.add(kb_algebra, kb_english, kb_biology, kb_geogr, kb_geoma, kb_chimiya, kb_informat, kb_litra, kb_nemec, kb_obshestvo, kb_history, kb_russian, kb_techno, kb_phisic)
        return await message.reply('🛍 Выберите *школьный предмет* из списка ниже', parse_mode= "Markdown", reply_markup=kb)



async def check_dz(call: types.CallbackQuery, name, db_name): # объект CallbackQuery, имя в дательном падеже, запись в БД
    c.execute(f"SELECT dz, author, time FROM domashka WHERE predmet = '{db_name}'")
    data = c.fetchone()

    if data[0] is not None:
        return await bot.answer_callback_query(call.id, f'\t\t📔 Доска д/з по {name}\n\n😇 Автор: @{data[1]}\n🕓 Время добавления: {data[2]}\n\n📚 Задание:\n{data[0]}', show_alert=True)
            
    else:
        return await bot.answer_callback_query(call.id, f"📔 Домашка по {name} отсутствует!", show_alert=True)
        

@dp.callback_query_handler()
async def callback_check_dz(call: types.CallbackQuery):
    if call.from_user.id != call.message.reply_to_message.from_user.id:
        return await bot.answer_callback_query(call.id, f"Вы не можете использовать чужие сообщения!") 

    if call.data.startswith('check'):
        if call.data == 'check_algebra':
            return await check_dz(call, 'алгебре', 'alg')

        elif call.data == 'check_enboys':
            return await check_dz(call, 'английскому языку(Иванова)', 'en1')
            
        elif call.data == 'check_engirls':
            return await check_dz(call, 'английскому языку(Хуснуллина)', 'en2')
        
        elif call.data == 'check_biology':
            return await check_dz(call, 'биологии', 'bio')

        elif call.data == 'check_geogr':
            return await check_dz(call, 'географии', 'geog')
        
        elif call.data == 'check_geoma':
            return await check_dz(call, 'геометрии', 'geom')
        
        elif call.data == 'check_chimiya':
            return await check_dz(call, 'химии', 'chem')

        elif call.data == 'check_iboys':
            return await check_dz(call, 'информатике(Санников)', 'info1')
            
        elif call.data == 'check_igirls':
            return await check_dz(call, 'информатике(Андреева)', 'info2')
        
        elif call.data == 'check_litra':
            return await check_dz(call, 'литературе', 'litra')

        elif call.data == 'check_nboys':
            return await check_dz(call, 'немецкому языку(Ильина)', 'nem1')
            
        elif call.data == 'check_ngirls':
            return await check_dz(call, 'немецкому языку(Разина)', 'nem2')

        elif call.data == 'check_obshestvo':
            return await check_dz(call, 'обществознанию', 'obchest')
        
        elif call.data == 'check_history':
            return await check_dz(call, 'истории', 'history')
        
        elif call.data == 'check_russian':
            return await check_dz(call, 'русскому языку', 'russian')

        elif call.data == 'check_tboys':
            return await check_dz(call, 'технологии(Мальчики)', 'techn1')
            
        elif call.data == 'check_tgirls':
            return await check_dz(call, 'технологии(Девочки)', 'techn2')

        elif call.data == 'check_phisic':
            return await check_dz(call, 'физике', 'phisic')

        
        elif call.data == 'check_techno':
            kb = types.InlineKeyboardMarkup()
            kb_tboys = types.InlineKeyboardButton(text='Мальчики', callback_data='check_tboys')
            kb_tgirls = types.InlineKeyboardButton(text='Девочки', callback_data='check_tgirls')
            kb_return = types.InlineKeyboardButton(text='Назад', callback_data='return')
            kb.add(kb_tboys, kb_tgirls, kb_return)
            return await call.message.edit_reply_markup(reply_markup=kb)

        elif call.data == 'check_nemec':
            kb = types.InlineKeyboardMarkup()
            kb_nboys = types.InlineKeyboardButton(text='Молодая (Ильина)', callback_data='check_nboys')
            kb_ngirls = types.InlineKeyboardButton(text='Старая (Разина)', callback_data='check_ngirls')
            kb_return = types.InlineKeyboardButton(text='Назад', callback_data='return')
            kb.add(kb_nboys, kb_ngirls, kb_return)
            return await call.message.edit_reply_markup(reply_markup=kb)

        elif call.data == 'check_informat':
            kb = types.InlineKeyboardMarkup()
            kb_iboys = types.InlineKeyboardButton(text='Санников', callback_data='check_iboys')
            kb_igirls = types.InlineKeyboardButton(text='Андреева', callback_data='check_igirls')
            kb_return = types.InlineKeyboardButton(text='Назад', callback_data='return')
            kb.add(kb_iboys, kb_igirls, kb_return)
            return await call.message.edit_reply_markup(reply_markup=kb)

        elif call.data == 'check_english':
            kb = types.InlineKeyboardMarkup()
            kb_enboys = types.InlineKeyboardButton(text='Иванова', callback_data='check_enboys')
            kb_engirls = types.InlineKeyboardButton(text='Хуснуллина', callback_data='check_engirls')
            kb_return = types.InlineKeyboardButton(text='Назад', callback_data='return')
            kb.add(kb_enboys, kb_engirls, kb_return)
            return await call.message.edit_reply_markup(reply_markup=kb)

    
    elif call.data == 'return':
        kb = types.InlineKeyboardMarkup()
        kb_algebra = types.InlineKeyboardButton(text='Алгебра', callback_data='check_algebra')
        kb_english = types.InlineKeyboardButton(text='Английский язык', callback_data='check_english')
        kb_biology = types.InlineKeyboardButton(text='Биология', callback_data='check_biology')
        kb_geogr = types.InlineKeyboardButton(text='География', callback_data='check_geogr')
        kb_geoma = types.InlineKeyboardButton(text='Геометрия', callback_data='check_geoma')
        kb_chimiya = types.InlineKeyboardButton(text='Химия', callback_data='check_chimiya')
        kb_informat = types.InlineKeyboardButton(text='Информатика', callback_data='check_informat')
        kb_litra = types.InlineKeyboardButton(text='Литература', callback_data='check_litra')
        kb_nemec = types.InlineKeyboardButton(text='Немецкий язык', callback_data='check_nemec')
        kb_obshestvo = types.InlineKeyboardButton(text='Обществознание', callback_data='check_obshestvo')
        kb_history = types.InlineKeyboardButton(text='История', callback_data='check_history')
        kb_russian = types.InlineKeyboardButton(text='Русский язык', callback_data='check_russian')
        kb_techno = types.InlineKeyboardButton(text='Технология', callback_data='check_techno')
        kb_phisic = types.InlineKeyboardButton(text='Физика', callback_data='check_phisic')

        kb.add(kb_algebra, kb_english, kb_biology, kb_geogr, kb_geoma, kb_chimiya, kb_informat, kb_litra, kb_nemec, kb_obshestvo, kb_history, kb_russian, kb_techno, kb_phisic)
        return await call.message.edit_reply_markup(reply_markup=kb)


@dp.message_handler(commands=['dz_new'])
async def new_dz(message: types.Message):
    if message.chat.type != 'private':
        return await message.reply("❌ Это можно использовать только в личке бота!")

    user_id = message.from_user.id
    c.execute(f"SELECT dostup FROM users WHERE id = {user_id}")
    data = c.fetchone()

    if int(data[0]) < 2 or data[0] is None:
        return await message.reply("У вас меньше 2 уровня доступа!")

    info = message.get_args().split(' ', maxsplit=1)

    if len(info) <= 1:
        return await message.reply("<b>💻 Для добавления домашки используйте следующий синтакисис:\n/dz_new *предмет* *домашка*\n\nПример:\n/dz_new en1 1. Учебник страница 51 упражнение устно, рабочая тетрадь стр 31 полностью\n\nДоступные предметы:\n\
            - alg - алгбера\n\
            - en1 - первая группа английского\n\
            - en2 - вторая группа английского\n\
            - bio - биология\n\
            - geom - геометрия\n\
            - geog - геометрия\n\
            - chem - химия\n\
            - info1 - первая группа информатики\n\
            - info2 - вторая группа информатики\n\
            - litra - литература\n\
            - nem1 - первая группа немецкого\n\
            - nem2 - вторая группа немецкого\n\
            - obchest - обществознание\n\
            - history - история\n\
            - russian - русский язык\n\
            - techn1 - первая группа технологии\n\
            - techn2 - вторая группа технологии\n\
            - phisic - физика</b>")

    if info[0] not in predmets:
        return await message.reply("<b>❗️ Вы указали неверный код предмета!\n\nДоступные предметы:\n\
            - alg - алгбера\n\
            - en1 - первая группа английского\n\
            - en2 - вторая группа английского\n\
            - bio - биология\n\
            - geom - геометрия\n\
            - geog - геометрия\n\
            - chem - химия\n\
            - info1 - первая группа информатики\n\
            - info2 - вторая группа информатики\n\
            - litra - литература\n\
            - nem1 - первая группа немецкого\n\
            - nem2 - вторая группа немецкого\n\
            - obchest - обществознание\n\
            - history - история\n\
            - russian - русский язык\n\
            - techn1 - первая группа технологии\n\
            - techn2 - вторая группа технологии\n\
            - phisic - физика</b>")


    c.execute(f"INSERT INTO logs(user_id, action, result) VALUES({user_id}, 'изменение д/з по {info[0]}', '{info[1]}')")
    db.commit()
    c.execute(f"UPDATE domashka SET dz = '{info[1]}', author = '{message.from_user.username}', time = '{datetime.datetime.now().strftime('%d.%m.%Y %H:%M:%S')}' WHERE predmet = '{info[0]}'")
    db.commit()
    await bot.send_message(logs_group, f"🤨 Замечена активность!\n\nID пользователя: <b>{user_id}</b>\nДействие: <b>изменение д/з по {info[0]}</b>\nРезультат: <b>{info[1]}</b>")
    return await message.reply(f"☑️ Вы успешно изменили домашку по <b>{predmets[info[0]]}</b> на:\n\n<b>{info[1]}</b>")

@dp.message_handler(commands=['dz_help'])
async def help_cmd(message: types.Message):
    user_id = message.from_user.id
    c.execute(f"SELECT dostup FROM users WHERE id = {user_id}")
    data = c.fetchone()

    if int(data[0]) < 1 or data[0] is None:
        return await message.reply("У вас меньше 1 уровня доступа!")

    return await message.reply("😍Команды ДомашкоБота:\n\n<b>/dz</b> - узнать домашнее задание(1+ уровень доступа)\n<b>/dz_help</b> - открыть меню помощи(1+ уровень доступа)\n<b>/dz_new *код предмета* *домашнее задание*</b> - обновить домашнее задание(2+ уровень доступа)\n<b>/dz_clear *код предмета*</b> - очистить домашнее задание(2+ уровень доступа)\n/dz_setadmin *id пользователя* *уровень доступа*")


@dp.message_handler(commands=['dz_clear'])
async def clear_cmd(message: types.Message):
    user_id = message.from_user.id
    c.execute(f"SELECT dostup FROM users WHERE id = {user_id}")
    data = c.fetchone()

    if int(data[0]) < 2 or data[0] is None:
        return await message.reply("У вас меньше 2 уровня доступа!")

    info = message.get_args()

    if len(info) <= 2:
        return await message.reply("<b>💻 Для обнуления домашки используйте следующий синтакисис:\n/dz_clear *предмет*\n\nПример:\n/dz_clear en1\n\nДоступные предметы:\n\
            - alg - алгбера\n\
            - en1 - первая группа английского\n\
            - en2 - вторая группа английского\n\
            - bio - биология\n\
            - geom - геометрия\n\
            - geog - геометрия\n\
            - chem - химия\n\
            - info1 - первая группа информатики\n\
            - info2 - вторая группа информатики\n\
            - litra - литература\n\
            - nem1 - первая группа немецкого\n\
            - nem2 - вторая группа немецкого\n\
            - obchest - обществознание\n\
            - history - история\n\
            - russian - русский язык\n\
            - techn1 - первая группа технологии\n\
            - techn2 - вторая группа технологии\n\
            - phisic - физика</b>")

    if info not in predmets:
        return await message.reply("<b>❗️ Вы указали неверный код предмета!\n\nДоступные предметы:\n\
            - alg - алгбера\n\
            - en1 - первая группа английского\n\
            - en2 - вторая группа английского\n\
            - bio - биология\n\
            - geom - геометрия\n\
            - geog - геометрия\n\
            - chem - химия\n\
            - info1 - первая группа информатики\n\
            - info2 - вторая группа информатики\n\
            - litra - литература\n\
            - nem1 - первая группа немецкого\n\
            - nem2 - вторая группа немецкого\n\
            - obchest - обществознание\n\
            - history - история\n\
            - russian - русский язык\n\
            - techn1 - первая группа технологии\n\
            - techn2 - вторая группа технологии\n\
            - phisic - физика</b>")


    c.execute(f"UPDATE domashka SET dz = null, author = null, time = null WHERE predmet = '{info}'")
    db.commit()
    c.execute(f"INSERT INTO logs(user_id, action, result) VALUES({user_id}, 'обнуление {info}', 'успешно')")
    db.commit()
    await bot.send_message(logs_group, f"🤨 Замечена активность!\n\nID пользователя: <b>{user_id}</b>\nДействие: <b>обнуление {info}</b>\nРезультат: <b>успешно</b>")
    return await message.reply(f"☑️ Вы обнулили домашку по <b>{predmets[info]}</b>")


@dp.message_handler(commands=['dz_setadmin'])
async def clear_cmd(message: types.Message):
    user_id = message.from_user.id
    c.execute(f"SELECT dostup FROM users WHERE id = {user_id}")
    data = c.fetchone()

    if int(data[0]) < 3 or data[0] is None:
        return await message.reply("❗️ У вас меньше 3 уровня доступа!")

    info = message.get_args().split(maxsplit=1)

    if len(info) < 2:
        return await message.reply("❗️ Неверный синтаксис!\n\nИспользуйте: <b>/dz_setadmin id lvl</b>\nПример: <b>/dz_setadmin 1231231234 2</b>")

    c.execute(f"SELECT id FROM users WHERE id = {info[0]}")
    data = c.fetchone()

    if data is None:
        return await message.reply("❗️ Пользователь не зарегистрирован в системе бота!")

    try:
        if int(info[1]) > -1 and int(info[1]) <= 3:
            c.execute(f"UPDATE users SET dostup = {info[1]} WHERE id = '{info[0]}'")
            db.commit()
            return await message.reply(f"❗️ Вы успешно назначили пользователю с ID <b>{info[0]}</b> доступ уровня <b>{info[1]}</b>!")

        else:
            return await message.reply("❗️ Уровень должен быть больше или равен 0, но меньше или равен 3!")
    
    except ValueError:
        return await message.reply("❗️ В аргументах команды нельзя использовать буквы!")


if __name__ == "__main__":
    executor.start_polling(dp, skip_updates=True)
